mcp/streamable.go

```diff
@@ -654,120 +701,370 @@ func (s *streamableClientConn) Read(ctx context.Context) (JSONRPCMessage, error)
+		return "", fmt.Errorf("initial POST request did not return an Mcp-Session-Id")
+	}
+	if newSessionID == "" {
+		// If the server didn't explicitly send a new one, assume the existing one is still valid.
+		newSessionID = currentSessionID
 	}
 
-	sessionID = resp.Header.Get("Mcp-Session-Id")
 	if resp.Header.Get("Content-Type") == "text/event-stream" {
 		go s.handleSSE(resp)
 	} else {
 		resp.Body.Close()
 	}
-	return sessionID, nil
+
+	return newSessionID, nil
 }
 
-func (s *streamableClientConn) handleSSE(resp *http.Response) {
-	defer resp.Body.Close()
+// startEventStreamReceiver continuously attempts to establish and maintain
+// the hanging GET connection for receiving Server-Sent Events (SSE).
+func (s *streamableClientConn) startEventStreamReceiver() {
+	backoffDuration := s.initialBackoff
+	retries := 0
 
-	done := make(chan struct{})
-	go func() {
-		defer close(done)
-		for evt, err := range scanEvents(resp.Body) {
-			if err != nil {
-				// TODO: surface this error; possibly break the stream
-				return
+	for {
+		select {
+		case <-s.done:
+			return // Connection is closed.
+		default:
+			// Continue
+		}
+
+		sessionID := s.sessionID.Load().(string)
+		if sessionID == "" {
+			// Session ID not yet established (first POST hasn't completed).
+			// Wait and retry.
+			time.Sleep(100 * time.Millisecond) // Avoid busy-waiting
+			continue
+		}
+
+		// Create a context for the current hanging GET request.
+		ctx, cancel := context.WithCancel(context.Background())
+		s.mu.Lock()
+		s.cancelHangingGet = cancel  // Store cancel function to allow external cancellation
+		lastEventID := s.lastEventID // Get the last processed event ID for replay
+		s.mu.Unlock()
+
+		// Perform the hanging GET request
+		err := s.performHangingGet(ctx, sessionID, lastEventID)
+
+		// Clean up after the hanging GET attempt
+		s.mu.Lock()
+		s.cancelHangingGet = nil // Clear the cancel function
+		s.mu.Unlock()
+		cancel() // Ensure the context for this specific GET is cancelled
+
+		if err == nil {
+			// Successful hanging GET, reset retry state
+			retries = 0
+			backoffDuration = s.initialBackoff
+			// Loop immediately to re-establish connection if it closed gracefully
+			continue
+		}
+
+		// Error occurred during hanging GET, check for retry
+		if retries >= s.maxRetries {
+			s.mu.Lock()
+			s.err = fmt.Errorf("failed to maintain SSE connection after %d retries: %w", s.maxRetries, err)
+			s.mu.Unlock()
+			s.Close() // Close the connection if persistent failure
+			return
+		}
+
+		// Apply exponential backoff with jitter
+		delay := backoffDuration + time.Duration(s.randSource.Int63n(int64(backoffDuration/2)))
+		select {
+		case <-s.done:
+			return // Connection closed during backoff
+		case <-time.After(delay):
+			retries++
+			backoffDuration *= 2                  // Exponential increase
+			if backoffDuration > 30*time.Second { // Cap backoff duration
```

Comment:
Why are this backoff delay computations so different between the client and server sides? Can they be pulled out into something more general?